File: transfer_wnd2.py
"""Transfer WND on Westend Asset Hub — with proper gas estimation."""

from web3 import Web3

RPC = "https://westend-asset-hub-eth-rpc.polkadot.io"
w3 = Web3(Web3.HTTPProvider(RPC))

MINER_KEY = "2191a9482334c6dcd94d58e5a0d929037c77b990d8a259fbb009a9812305da72"
MINER_ADDR = "0x54C375b935344fC3c16E5e83F444DB54dE0E6eB9"
DEPLOYER_ADDR = "0x350EB21005e911f4493a93449DDD329dE1fd7964"
VALIDATOR_ADDR = "0x7453dc226c58C5758eDd9De79162Fe2aeFDf229D"

chain_id = w3.eth.chain_id
print(f"Chain ID: {chain_id}")
print(f"Gas price: {w3.eth.gas_price}")

miner_bal = w3.eth.get_balance(MINER_ADDR)
print(f"Miner balance: {Web3.from_wei(miner_bal, 'ether')} WND ({miner_bal} wei)")

nonce = w3.eth.get_transaction_count(MINER_ADDR)
print(f"Miner nonce: {nonce}")

# Try estimating gas for a simple transfer
try:
    est = w3.eth.estimate_gas(
        {
            "from": MINER_ADDR,
            "to": DEPLOYER_ADDR,
            "value": Web3.to_wei(1, "ether"),
        }
    )
    print(f"Estimated gas: {est}")
except Exception as e:
    print(f"Gas estimation error: {e}")

# Try with EIP-1559 (type 2 tx)
try:
    base_fee = w3.eth.get_block("latest").get("baseFeePerGas", 0)
    print(f"Base fee: {base_fee}")
except Exception as e:
    print(f"No base fee: {e}")

# Attempt transfer with type 2 tx
print("\n--- Attempting transfer ---")
try:
    tx1 = {
        "from": MINER_ADDR,
        "to": DEPLOYER_ADDR,
        "value": Web3.to_wei(6, "ether"),
        "nonce": nonce,
        "chainId": chain_id,
        "type": 2,
        "maxFeePerGas": w3.eth.gas_price * 2,
        "maxPriorityFeePerGas": w3.eth.gas_price,
        "gas": 21000,
    }
    signed1 = w3.eth.account.sign_transaction(tx1, MINER_KEY)
    tx_hash1 = w3.eth.send_raw_transaction(signed1.raw_transaction)
    receipt1 = w3.eth.wait_for_transaction_receipt(tx_hash1, timeout=120)
    print(f"✅ Miner → Deployer: 6 WND | TX: {receipt1.transactionHash.hex()}")
except Exception as e:
    print(f"Type 2 failed: {e}")
    # Try legacy tx with explicit gas price
    try:
        gas_price = max(w3.eth.gas_price, 1_000_000_000)  # min 1 Gwei
        tx1 = {
            "to": DEPLOYER_ADDR,
            "value": Web3.to_wei(6, "ether"),
            "gas": 21000,
            "gasPrice": gas_price,
            "nonce": nonce,
            "chainId": chain_id,
        }
        signed1 = w3.eth.account.sign_transaction(tx1, MINER_KEY)
        tx_hash1 = w3.eth.send_raw_transaction(signed1.raw_transaction)
        receipt1 = w3.eth.wait_for_transaction_receipt(tx_hash1, timeout=120)
        print(f"✅ Miner → Deployer: 6 WND (legacy) | TX: {receipt1.transactionHash.hex()}")
    except Exception as e2:
        print(f"Legacy also failed: {e2}")

File: luxtensor/contracts/scripts/transfer-wnd2.js
/**
 * Transfer WND on Westend Asset Hub — EIP-1559 with explicit params
 */
const hre = require("hardhat");

async function main() {
    const provider = new hre.ethers.JsonRpcProvider(
        "https://westend-asset-hub-eth-rpc.polkadot.io",
        { chainId: 420420421, name: "westend" }
    );

    const minerKey = "0x2191a9482334c6dcd94d58e5a0d929037c77b990d8a259fbb009a9812305da72";
    const miner = new hre.ethers.Wallet(minerKey, provider);

    const deployer = "0x350EB21005e911f4493a93449DDD329dE1fd7964";
    const validator = "0x7453dc226c58C5758eDd9De79162Fe2aeFDf229D";

    const bal = await provider.getBalance(miner.address);
    console.log("Miner balance:", hre.ethers.formatEther(bal), "WND");
    console.log("Nonce:", await provider.getTransactionCount(miner.address));

    // Get latest block for base fee
    const block = await provider.getBlock("latest");
    console.log("Latest block:", block.number, "baseFeePerGas:", block.baseFeePerGas?.toString());

    // Approach 1: type 2 EIP-1559 with explicit gasLimit
    console.log("\n--- Approach 1: EIP-1559 type 2 ---");
    try {
        const nonce = await provider.getTransactionCount(miner.address);
        const tx = {
            type: 2,
            to: deployer,
            value: hre.ethers.parseEther("6"),
            chainId: 420420421,
            nonce: nonce,
            maxFeePerGas: 220000000n,
            maxPriorityFeePerGas: 20000000n,
            gasLimit: 21000n,
        };
        const signed = await miner.signTransaction(tx);
        console.log("Signed TX (first 100 chars):", signed.substring(0, 100));
        const response = await provider.broadcastTransaction(signed);
        console.log("TX hash:", response.hash);
        const receipt = await response.wait();
        console.log("✅ Success! Gas used:", receipt.gasUsed.toString());
    } catch (e) {
        console.error("❌ Approach 1 failed:", e.message?.substring(0, 300) || e);
    }

    // Approach 2: type 0 legacy with exact gas price
    console.log("\n--- Approach 2: Legacy type 0 ---");
    try {
        const nonce = await provider.getTransactionCount(miner.address);
        const tx = {
            type: 0,
            to: deployer,
            value: hre.ethers.parseEther("6"),
            chainId: 420420421,
            nonce: nonce,
            gasPrice: 100000000n,
            gasLimit: 21000n,
        };
        const signed = await miner.signTransaction(tx);
        const response = await provider.broadcastTransaction(signed);
        console.log("TX hash:", response.hash);
        const receipt = await response.wait();
        console.log("✅ Success! Gas used:", receipt.gasUsed.toString());
    } catch (e) {
        console.error("❌ Approach 2 failed:", e.message?.substring(0, 300) || e);
    }

    // Approach 3: type 2 with higher gas limit
    console.log("\n--- Approach 3: EIP-1559 high gas ---");
    try {
        const nonce = await provider.getTransactionCount(miner.address);
        const tx = {
            type: 2,
            to: deployer,
            value: hre.ethers.parseEther("6"),
            chainId: 420420421,
            nonce: nonce,
            maxFeePerGas: 500000000n,
            maxPriorityFeePerGas: 100000000n,
            gasLimit: 500000n,
        };
        const signed = await miner.signTransaction(tx);
        const response = await provider.broadcastTransaction(signed);
        console.log("TX hash:", response.hash);
        const receipt = await response.wait();
        console.log("✅ Success! Gas used:", receipt.gasUsed.toString());
    } catch (e) {
        console.error("❌ Approach 3 failed:", e.message?.substring(0, 300) || e);
    }

    // Final balances
    console.log("\n=== Final Balances ===");
    console.log("Deployer: ", hre.ethers.formatEther(await provider.getBalance(deployer)), "WND");
    console.log("Validator:", hre.ethers.formatEther(await provider.getBalance(validator)), "WND");
    console.log("Miner:    ", hre.ethers.formatEther(await provider.getBalance(miner.address)), "WND");
}

main().catch(console.error);
